File: mobile/ios/PaymentSDKPlugin.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(PaymentSDKPlugin, RCTEventEmitter)

RCT_EXTERN_METHOD(execute:(NSString *)action args:(NSString *)args)

@end

File: mobile/ios/PaymentSDKPlugin.swift
import CommonCrypto
import Foundation
import PaymentSDK
import React

@objc(PaymentSDKPlugin)
final class PaymentSDKPlugin: RCTEventEmitter {
    
    private lazy var actionHandler: ActionHandler = .init(context: self)
    private(set) var hasListeners: Bool = false
    
    @objc(execute:args:)
    func execute(action: String, args: String) {
        actionHandler.execute(
            action: action,
            args: args
        )
    }
    
    override func startObserving() {
        hasListeners = true
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    override func supportedEvents() -> [String]! {
        ActionEnum.allCases.map {
            $0.rawValue
        }
    }
    
    override class func requiresMainQueueSetup() -> Bool {
        true
    }
}
